### ISSUE TITLE
TemplateResult returns incorrect information

### ISSUE THREAD

**Issue by ekr3peeK** [select2/select2#5871]
I noticed this bug when I was trying to hide already selected options from appearing in the dropdown of select2.

The `templateResult` callback function, provides a `data` variable, which holds information about the `option` that will be rendered. Between this information, a property appears with the name of `selected` and a `true/false` value.

I was presuming, that this represents if the option iterated is currently selected or not, but it seems that once this property is set to true, it will stay true, even when the user selects a different option. You can try this out in the JSBin, and check the logs that I output on the screen.
[
JSBIN](https://jsbin.com/nezoqawujo/1/edit?html,js,output)

**Steps to reproduce in JSBIN:**

1. Open the Select2, and verify the logs: it states that the option with the id of 1 is currently selected, which is true
2. Select the second option from the dropdown
3. Open the Select2 once again, and verify the logs: it states that both the option with the id of 1 is selected, and the option with the id of 2.

With a bit of modification (ie. by adding the property `multiple: true` at initialization), it seems that with a multiple select, the property is updated normally. It seems that this bug appears because when the selection changes in a single select, the property is not explicitly set to false to every previously selected element, like it is set to false, when the client explicitly deselects something from the list.

**WORKAROUND:**

For a workaround, at the moment, I am hiding the already selected options, via css, with the following line:

    .select2-container .select2-dropdown .select2-results ul.select2-results__options li[aria-selected=true] {
        display: none;
    }

**Comment by kevin-brown** [comment]
> I was presuming, that this represents if the option iterated is currently selected or not, but it seems that once this property is set to true, it will stay true, even when the user selects a different option.

You know what they say about assumptions...

https://github.com/select2/select2/blob/defe232bf6d233f17f3c26b776ad179ad26fe2bc/src/js/select2/results.js#L138-L139

The `selected` property is not the source of truth for Select2 and only exists for internal compatibility with AJAX requests. The source of truth for Select2 is the `<option>` element itself, which can be accessed by `data.element` if needed.

> For a workaround, at the moment, I am hiding the already selected options, via css, with the following line:

This will break in Select2 4.1.0 as we are moving towards an actual class which will indicate if an option is selected or not.

**Comment by ekr3peeK** [comment]
Still, based on how the multiple select behaves with the same property, I would at least consider describing this in the options, for other developers to not fall in the same pit as I did.

Because what I did, was to implement the `templateResult` for a multiple select (on which this works with the `selected` property), and later found my bug when using the same `templateResult` for a single select box.

Based on your standing point, what is now the correct way, to completely hide (instead of highlight) already selected elements in the results dropdown?

Am I correct, that to alter the results in the dropdown list, the only javascript functionality one can use is the `templateResults` property?
